refactor(preprocessors): replace hardcoded values with config/CLI par…

Vedavyasa21 opened the pull request:
### **User description**
Summary
Replaced hardcoded values in the preprocessors with configurable constants or CLI parameters to improve maintainability and flexibility.

Changes Made
Listed and replaced hardcoded numeric values, paths, and parameters in the processor files.
Pulled configuration values from a central config or CLI input. Verified functionality remains intact.


___

### **PR Type**
Enhancement


___

### **Description**
- Created centralized constants module for image processing parameters

- Replaced hardcoded values across preprocessors with configurable constants

- Improved code maintainability by consolidating magic numbers into named constants

- Enhanced flexibility for tuning image processing parameters


___

### Diagram Walkthrough


```mermaid
flowchart LR
  A["Hardcoded Values<br/>in Processors"] -->|"Extract & Centralize"| B["Constants Module<br/>image_processing.py"]
  B -->|"Import Constants"| C["CropPage.py"]
  B -->|"Import Constants"| D["CropOnMarkers.py"]
  B -->|"Import Constants"| E["FeatureBasedAlignment.py"]
  B -->|"Import Constants"| F["builtins.py"]
  C -->|"Use Constants"| G["Improved Maintainability"]
  D -->|"Use Constants"| G
  E -->|"Use Constants"| G
  F -->|"Use Constants"| G
```



<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>__init__.py</strong><dd><code>Constants package initialization and exports</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/constants/__init__.py

<ul><li>Created new constants package initialization file<br> <li> Exports all image processing constants for easy access<br> <li> Provides centralized entry point for constants module</ul>


</details>


  </td>
  <td><a href="https://github.com/Udayraj123/OMRChecker/pull/250/files#diff-5ab8144b53a30871714ebf02efadae918b888063bc1a52c1a2318b4258070ddc">+22/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>image_processing.py</strong><dd><code>Centralized image processing constants definition</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/constants/image_processing.py

<ul><li>Defined 21 image processing constants organized by functionality<br> <li> Includes parameters for blur operations, thresholding, edge detection, <br>and erosion<br> <li> Covers color values, kernel sizes, and algorithm-specific parameters<br> <li> Replaces previously hardcoded magic numbers throughout codebase</ul>


</details>


  </td>
  <td><a href="https://github.com/Udayraj123/OMRChecker/pull/250/files#diff-43d0eee6cf3d4781bc591ca0fd21f7f4bc2492269857706ada8c5990b91e3d3b">+61/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>CropOnMarkers.py</strong><dd><code>Replace hardcoded values with imported constants</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/processors/CropOnMarkers.py

<ul><li>Imported 12 constants from centralized constants module<br> <li> Replaced hardcoded erosion kernel size and iterations with <br><code>EROSION_PARAMS</code><br> <li> Replaced hardcoded quadrant division values with <code>QUADRANT_DIVISION</code> <br>constants<br> <li> Replaced hardcoded color values with named constants like <br><code>DEFAULT_WHITE_COLOR</code>, <code>MARKER_RECTANGLE_COLOR</code><br> <li> Updated Gaussian blur and normalization to use centralized parameters</ul>


</details>


  </td>
  <td><a href="https://github.com/Udayraj123/OMRChecker/pull/250/files#diff-bd84f46fea3f3e6e963289c20cc04750395653d7b4deee131e2a2c2d2b802233">+30/-10</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>CropPage.py</strong><dd><code>Migrate page cropping hardcoded values to constants</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/processors/CropPage.py

<ul><li>Imported 10 constants for page cropping operations<br> <li> Replaced <code>MIN_PAGE_AREA</code> hardcoded value with <code>MIN_PAGE_AREA_THRESHOLD</code><br> <li> Replaced hardcoded Canny edge detection thresholds with <br><code>CANNY_EDGE_PARAMS</code><br> <li> Replaced hardcoded threshold parameters with <code>PAGE_THRESHOLD_PARAMS</code><br> <li> Updated contour drawing colors and line widths to use named constants</ul>


</details>


  </td>
  <td><a href="https://github.com/Udayraj123/OMRChecker/pull/250/files#diff-56714886cca581e65e073e869f93b22e3badc3276dceadebac65afa18fc44d7e">+21/-10</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>FeatureBasedAlignment.py</strong><dd><code>Use constants for feature alignment defaults</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/processors/FeatureBasedAlignment.py

<ul><li>Imported default feature detection constants<br> <li> Replaced hardcoded <code>maxFeatures</code> default value with <code>DEFAULT_MAX_FEATURES</code><br> <li> Replaced hardcoded <code>goodMatchPercent</code> default with <br><code>DEFAULT_GOOD_MATCH_PERCENT</code></ul>


</details>


  </td>
  <td><a href="https://github.com/Udayraj123/OMRChecker/pull/250/files#diff-108d2c78fd9292a51bcadd0ac64bed80ef157c4f51641e0a47a03df612dc94c5">+6/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>builtins.py</strong><dd><code>Apply constants to builtin blur processors</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/processors/builtins.py

<ul><li>Imported median blur and Gaussian blur constants<br> <li> Replaced hardcoded median blur kernel size default with <br><code>DEFAULT_MEDIAN_BLUR_KERNEL_SIZE</code><br> <li> Replaced hardcoded Gaussian blur kernel size and sigma defaults with <br><code>DEFAULT_GAUSSIAN_BLUR_PARAMS</code></ul>


</details>


  </td>
  <td><a href="https://github.com/Udayraj123/OMRChecker/pull/250/files#diff-406e88b67edd596ef32df5c9d5fbd219bb97fb6c612901b2e80cc6914bfb0254">+7/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

Udayraj123 commented on src/constants/image_processing.py:
```diff
@@ -0,0 +1,61 @@
+ERODE_RECT_COLOR = (50, 50, 50)
+NORMAL_RECT_COLOR = (155, 155, 155)
+EROSION_PARAMS = {
+    "kernel_size": 5,
```
can we make this a tuple (5,5) similar to DEFAULT_GAUSSIAN_BLUR_PARAMS?

Udayraj123 reviewed (commented):
thanks for the PR @veda1121, looks like a good progress towards removing the hardcodings.

Vedavyasa21 commented:
Yeah i have done the Changes
Renamed threshold1 → lower_threshold and threshold2 → upper_threshold across all references for better clarity.
Removed the unnecessary __all__ list from src/constants/__init__.py since it was restricting imports and wasn’t needed.
